### ISSUE TITLE
Fine tuning doesn't work well on my knowledge graph

### ISSUE THREAD

**Issue by jr011** [DeepGraphLearning/ULTRA#19]
Hello, This works brilliantly, but it doesn't work very well on my knowledge graph. My knowledge graph has over 300,000 data and 18 relationships. I trained it with one of its subgraphs with 30,000 data for 10 epochs with batchsize set to 32. At the end of the training it had a very small loss, but the MRR and Hit didn't improve compared to the test after the first epoch, what could possibly be the cause of this?

![image](https://github.com/DeepGraphLearning/ULTRA/assets/60546139/a2f0e911-81ba-4116-a576-709348c14b5c)

**Comment by migalkin** [comment]
Hi, 

did you make some modifications to the model? Because the original checkpoint has about 177k parameters but your log shows the model has 796k parameters.

Besides, a sudden loss drop from 0.10 to 0.003 might indicate there are some issues with preparing inverse edges and relation types - basically, the model discovers those shortcut edges instead of generalizing. Hard to say more without seeing the graph construction process. 

During training, we call a function [`remove_easy_edges`](https://github.com/DeepGraphLearning/ULTRA/blob/c414f83d9ec1e4726f396c24beab6311a2869f1a/ultra/base_nbfnet.py#L54) that, by default, removes an edge with a query edge type (`h,r,t`) and its inverse (`t, r_inv, h`) (inverse relation id must be `r_id + num_original_relations`, for example, in a graph with 18 relations, relation ID 20 must be the inverse of relation ID 2). You might want to add `remove_one_hop=True` option to the entity model config to remove all edges between head and tail nodes (not just the query edge type), but I suspect the main problem is in inverse edge construction in your graph.

**Comment by jr011** [comment]
My process of constructing the graph is as follows: the relationships in my graph are all one-way. I just use most of the triples in the graph as the training set, and a small part as the valid and test set, and ensure that the valid set and test The entities and relationships of the set have appeared in the training set.

Isn't the reverse edge construction automatically constructed by the code? I did not modify the code of datasets.py. Or do I need to manually process the dataset?

Besides, I posted the wrong picture, this is indeed my training process on the modified model, but the situation is the same in the training of the original model.

**Comment by migalkin** [comment]
Yes, if you are subclassing from the [`TransductiveDataset`](https://github.com/DeepGraphLearning/ULTRA/blob/c414f83d9ec1e4726f396c24beab6311a2869f1a/ultra/datasets.py#L240) then it should take care of the inverses.

Still, there might be many other one-hop edges between head and tail and model finds shortcuts among them, so I'd encourage you to try the `remove_one_hop=True` option in the config for the `entity_model`, eg:

```yaml
  entity_model:
    class: IndNBFNet
    input_dim: 64
    hidden_dims: [64, 64, 64, 64, 64, 64]
    message_func: distmult
    aggregate_func: sum
    short_cut: yes
    layer_norm: yes
    remove_one_hop: yes    # <- new line here
```

and run fine-tuning with this config

**Comment by jr011** [comment]
Thank you very much for your reply. I tried running the code with this config, but the situation is still the same. This is my knowledge graph structure and the config for this run. Can you help me analyze the possible reasons? Thank you again for your reply.
![image](https://github.com/DeepGraphLearning/ULTRA/assets/60546139/dc09f71c-1f33-47c5-aa69-ac7c87045ec9)
![image](https://github.com/DeepGraphLearning/ULTRA/assets/60546139/20c512b9-bec9-4a7a-b77c-61c748d334f2)

**Comment by migalkin** [comment]
Turns out this question is a duplicate of https://github.com/daniel4x/daniel4x.github.io/issues/1 with more details on the graph structure and model modifications. I's suggest to follow up there
